src/Corax/IndexWriter.cs

```diff
@@ -435,71 +449,93 @@ private IndexWriter(IndexFieldsMapping fieldsMapping)
-            _entriesContainerId = Transaction.OpenContainer(Constants.IndexWriter.EntriesContainerSlice);
-            _entryIdToOffset = Transaction.LookupFor<Int64LookupKey>(Constants.IndexWriter.EntryIdToOffsetSlice);
-
-            _indexMetadata = Transaction.CreateTree(Constants.IndexMetadataSlice);
-            _lastEntryId = _indexMetadata?.ReadInt64(Constants.IndexWriter.LastEntryIdSlice) ?? 0;
-            _initialNumberOfEntries = _indexMetadata?.ReadInt64(Constants.IndexWriter.NumberOfEntriesSlice) ?? 0;
+            Init();
+        }
 
-            _documentBoost = Transaction.FixedTreeFor(Constants.DocumentBoostSlice, sizeof(float));
+        public long Index(string key, ReadOnlySpan<byte> data, float documentBoost)
+        {
+            return Index(Encoding.UTF8.GetBytes(key), data, documentBoost);
         }
-        
-        public long Index(Span<byte> data, float documentBoost)
+
+        public long Index(ReadOnlySpan<byte> key, ReadOnlySpan<byte> data, float documentBoost)
         {
-            var idInTree = Index(data);
+            var idInTree = Index(key, data);
             AppendDocumentBoost(EntryIdEncodings.DecodeAndDiscardFrequency(idInTree), documentBoost);
             return idInTree;
         }
-        
+
+        public long Index(string key, ReadOnlySpan<byte> data)
+        {
+            return Index(Encoding.UTF8.GetBytes(key), data);
+        }
         /// <returns>Encoded id (with freq/container_type)</returns>
-        public unsafe long Index(Span<byte> data)
+        public long Index(ReadOnlySpan<byte> key, ReadOnlySpan<byte> data)
         {
             _numberOfModifications++;
-            // align to 16 bytes boundary to ensure that we have some (small) space for updating in-place entries
-            var entryContainerId = Container.Allocate(Transaction.LowLevelTransaction, _entriesContainerId, data.Length, out var space);
-            data.CopyTo(space);
-            space = space.Slice(data.Length);
-            space.Clear();// clean any old data that may have already been there
-
-            var context = Transaction.Allocator;
-            
             var entryId = ++_lastEntryId;
-            _entryIdToOffset.Add(entryId, entryContainerId);
+            _entriesAllocator.Allocate(key.Length, out var keyStr);
+            key.CopyTo(keyStr.ToSpan());
+            _entriesAllocator.Allocate(data.Length, out var dataStr);
+            data.CopyTo(dataStr.ToSpan());
+            _bufferedIndexedEntries[entryId] = dataStr;
 
+            ref var container = ref CollectionsMarshal.GetValueRefOrAddDefault(_indexedEntriesByKey, keyStr, out var exists);
+            if (exists == false)
+            {
+                container.SingleItem = entryId;
+            }
+            else
+            {
+                container.Items = new List<long>
+                {
+                    container.SingleItem, 
+                    entryId,
+                };
+            }
```

Comment:
It's possible to hit that block twice? It can override previous values then, right?

Reply:
You are correct, that is a problem then.